File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        for (int i = 0; i < isConnected.length; i++) {
            for (int j = i; j < isConnected[i].length; j++) {
                if (isConnected[i][j] == 1) {
                    graph.computeIfAbsent(i, k -> new HashSet<>()).add(j);
                    graph.computeIfAbsent(j, k -> new HashSet<>()).add(i);
                }
            }
        }
        Set<Integer> seen = new HashSet<>();
        int res = 0;
        for (Map.Entry<Integer, Set<Integer>> con : graph.entrySet()) {
            int v = con.getKey();
            if (!seen.contains(v)) {
                res++;
                Queue<Integer> q = new LinkedList<>();
                seen.add(v);
                q.add(v);
                while (!q.isEmpty()) {
                    int node = q.poll();
                    for (int nb : graph.get(node)) {
                        if (!seen.contains(nb)) {
                            seen.add(nb);
                            q.add(nb);
                        }
                    }
                }
            }
        }
        return res;
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        graph = defaultdict(list)
        for i in range(n):
            for j in range(i+1, n):
                if isConnected[i][j]:
                    graph[i].append(j)
                    graph[j].append(i)
        seen = set()
        res = 0
        for i in range(n):
            if i not in seen:
                res += 1
                q = deque([i])
                seen.add(i)
                while q:
                    v = q.popleft()
                    for nb in graph[v]:
                        if nb not in seen:
                            seen.add(nb)
                            q.append(nb)
        return res